KOGITO-6958: DashBuilder YAML Editor

jesuino opened the pull request:
Introduces a new editor for Dashbuilder in YML format. It uses monaco editor and is based on the existing serverless editor.

tiagobento commented:
@jesuino Hi William! Looks like there's some linting errors happening on the new package. Can you please take a look?

jesuino commented:
Hello @barboras7 

To run the editor first build in root directory the following command:

```
yarn build:prod:until @kie-tools/dashbuilder-editor
```

It will take almost one hour to run (if running for the first time). then you can start the editor:

```
cd packages/dashbuilder-editor
yarn start
```
Then access the editor at: `http://localhost:9007`

You can try any YAML from our guide and it should work: https://www.dashbuilder.org/docs/#chap-dashbuilder-yaml-guides

barboras7 commented:
Hi @jesuino, 
thanks for the information. My testing was actually in progress already, I am sharing my findings below (there might be some more, I still do have some scenarios in my head that I want to try :slightly_smiling_face: ). 

Also, I am aware that some of the issues I am sharing below are not directly caused by this PR but I see this as a good place to start the discussion. I can open a Jira for them if we agree on that.

1. There's a typo in the imports description, see attached screenshot (dashbuilderś instead of dashbuilders)
![Screenshot from 2022-04-25 14-26-01](https://user-images.githubusercontent.com/11525583/165088829-c7c45d82-eaad-4280-a0a3-fbf4cbe57791.png)
2. The editor width is set to default every time the user sets preview to true and then back to false (as a user, I want to have the editor width I set during my entire interaction with the editor)
3. If a user uploads a yaml file, it is displayed in the preview but the raw content is not available in the editor (which means that the user can not edit the yaml file to their needs)
![Screenshot from 2022-04-25 14-21-00](https://user-images.githubusercontent.com/11525583/165089418-1092a1ea-973f-4500-b74d-14109d00d500.png)
4. If a user is working with a large yaml file, it is difficult to find the issue with the content of the yaml file (it would be useful to add the number of the first incorrect row into the notification)
![Screenshot from 2022-04-25 14-15-53](https://user-images.githubusercontent.com/11525583/165089808-94d073b5-64f9-4626-ab9a-477cb3ed321d.png)
5. I prepared some HTML code in business central and tested it in Dashbuilder YAML editor. The code was not read properly as it is probably not supported. Should all HTML options from business central be supported in the YAML editor? 
![Screenshot from 2022-04-25 14-16-26](https://user-images.githubusercontent.com/11525583/165090041-b57def9d-6cf1-4ef9-ab97-ab250bfff58a.png)

Please, let me know what you think :slightly_smiling_face:

jesuino commented:
Hello Barbora,

THanks for the feedback. Responses inline:
>> There's a typo in the imports description, see attached screenshot (dashbuilderś instead of dashbuilders)

Thanks! We fixed the typo and started our work to have modes on dashbuilder runtime, where if Dashbuilder is in mode "editor" so it should not allow upload. We also added a default content so users can start their own dashboard from there:

https://drive.google.com/file/d/1zsQNJgo2n2146vgbjN2dM0weWQa9lMTM/view

>> The editor width is set to default every time the user sets preview to true and then back to false (as a user, I want to have the editor width I set during my entire interaction with the editor)

I completely agree with you, but this is tough because the resizable panel component does not hold the width information in the same format that it uses to set the initial width. Hence I don't think it is possible ATM. 


>> If a user uploads a yaml file, it is displayed in the preview but the raw content is not available in the editor (which means that the user can not edit the yaml file to their needs)

Uploaded should be disabled and we did it in my latest commit.

>> If a user is working with a large yaml file, it is difficult to find the issue with the content of the yaml file (it would be useful to add the number of the first incorrect row into the notification)


>> I prepared some HTML code in business central and tested it in Dashbuilder YAML editor. The code was not read properly as it is probably not supported. Should all HTML options from business central be supported in the YAML editor?

The editor should soon become not compatible with the same editor from BC. Regarding the HTML it does not work because some parts are lost during the YAML to JSON conversion and some reserved characters are not converted correctly. A way to workaround this is by using multiple line in YAML:

![image](https://user-images.githubusercontent.com/359820/165187186-b27d7b32-d684-43cc-9d4d-b61357af860a.png)

Here's the code I tried:
```
pages:
  - components:
      - html: >-
          <p style="color:red; font-size: xx-large">Hello World</p>
```

Please let me know if you have any question!

barboras7 commented:
Hi @jesuino,

thank you for your responses. I am adding one suggestion into the PR review, not necessarily needed to be fixed now. 

Also, you forgot to answer this concern from the first review:
`If a user is working with a large yaml file, it is difficult to find the issue with the content of the yaml file (it would be useful to add the number of the first incorrect row into the notification)
`
Would this be possible? Now, or in the future? 

Thank you,
Barbora

jesuino commented:
Hello @barboras7 

Sorry, I missed the point regarding the error message.

I sent a change which I was working on which was a new screen for error. Now, the error messed could be messed sometimes:
![image](https://user-images.githubusercontent.com/359820/165823477-aa8f3e09-1612-4be1-9466-d60e8a6f47fc.png)

this is something that will be solved when we (hopefully soon) solve with schema validation. See how it works with SWF editor.

![image](https://user-images.githubusercontent.com/359820/165823578-d71af8df-b39b-4c7c-a6fa-4a827e42bc3c.png)


Regarding not used properties, yes, we should focus on it. I opened a JIRA for leftovers[1], but still didn't have the time to focus on it.

Thanks!

[1] https://issues.redhat.com/browse/KOGITO-6882

barboras7 reviewed (approved):
Looks good, thanks :tada:
